src/type_conversion/runpython.c

```diff
@@ -1,165 +1,132 @@
-          .then(runInternal);
-      }
-    }
-    return new Promise(internal);
-  };
-});
+#define QUIT_IF_NZ(x)                                                          \
+  do {                                                                         \
+    if (x) {                                                                   \
+      goto fail;                                                               \
+    }                                                                          \
+  } while (0)
 
 int
-runpython_init_py()
+runpython_init()
 {
+  bool success = false;
+  JsRef pyodide_py_proxy = Js_ERROR;
+  JsRef globals_proxy = Js_ERROR;
+
+  // borrowed
   PyObject* builtins = PyImport_AddModule("builtins");
-  if (builtins == NULL) {
-    return 1;
-  }
+  QUIT_IF_NULL(builtins);
 
+  // borrowed
   PyObject* builtins_dict = PyModule_GetDict(builtins);
-  if (builtins_dict == NULL) {
-    return 1;
-  }
+  QUIT_IF_NULL(builtins_dict);
 
+  // borrowed
   PyObject* __main__ = PyImport_AddModule("__main__");
-  if (__main__ == NULL) {
-    return 1;
-  }
+  QUIT_IF_NULL(__main__);
 
+  // globals is static variable, borrowed
   globals = PyModule_GetDict(__main__);
-  if (globals == NULL) {
-    return 1;
-  }
+  QUIT_IF_NULL(globals);
+  Py_INCREF(globals); // to owned
 
-  if (PyDict_Update(globals, builtins_dict)) {
-    return 1;
-  }
+  QUIT_IF_NZ(PyDict_Update(globals, builtins_dict));
 
-  pyodide = PyImport_ImportModule("pyodide");
-  if (pyodide == NULL) {
-    return 1;
-  }
+  // pyodide_py is static variable, new
+  pyodide_py = PyImport_ImportModule("pyodide");
+  QUIT_IF_NULL(pyodide_py);
 
-  return 0;
-}
+  pyodide_py_proxy = python2js(pyodide_py);
+  if (pyodide_py_proxy == Js_ERROR) {
+    goto fail;
+  }
+  // Currently by default, python2js copies dicts into objects.
+  // We want to feed Module.globals back to `eval_code` in `pyodide.runPython`
+  // (see definition in pyodide.js) but because the round trip conversion
+  // py => js => py for a dict object is a JsProxy, that causes trouble.
+  // Instead we explicitly call pyproxy_new.
+  // We also had to add ad-hoc modifications to _pyproxy_get, etc to support
+  // this. I (HC) will fix this with the rest of the type conversions
+  // modifications.
+  Py_INCREF(globals); // pyproxy_new steals argument
+  globals_proxy = pyproxy_new(globals);
+  if (globals_proxy == Js_ERROR) {
+    goto fail;
+  }
+  QUIT_IF_NZ(runpython_init_js(pyodide_py_proxy, globals_proxy));
 
-EM_JS(int, runpython_finalize_js, (), {
-  Module.version = function()
-  {
-    Module.runPython("import pyodide");
-    return Module.runPython("pyodide.__version__");
-  };
   return 0;
-});
-
-int
-runpython_init()
-{
-  return runpython_init_js() || runpython_init_py() || runpython_finalize_js();
+fail:
+  if (PyErr_Occurred()) {
+    PyErr_Print();
+  }
+  Py_CLEAR(pyodide_py);
+  Py_CLEAR(globals);
+  hiwire_decref(pyodide_py_proxy);
+  hiwire_decref(globals_proxy);
+  return -1;
```

Comment:
Wouldn't +1 be a more classical exit status for errors?

Reply:
I'm trying to be consistent with the Python conventions as much as possible. The majority of Python functions that return ints have the convention that they `Return 0 on success or -1 on failure.` The one annoying exception is that the argument parsing functions return 1 on success and 0 on fail. There are also special cases for functions for which -1 is a valid return value, in that case you have to check `PyErr_Occurred()`.

Reply:
In particular, python module initialization functions have this calling convention.

Reply:
Right I guess the point is that if the function returns a boolean it's `1` = true, `0` = false, `-1` = failed. And there are a lot of functions that naturally return a nonnegative integer, like the length of a list.